pkg/controller/storagecluster/reconcile.go

```diff
@@ -744,11 +745,33 @@ func newCephCluster(sc *ocsv1.StorageCluster, cephImage string) *cephv1.CephClus
 		},
 	}
 
-	if len(sc.Spec.MonDataDirHostPath) != 0 {
+	// If both `MonDataDirHostPath` and `monPVCTemplate` is provided, the mons will
+	// be provisioned on the provided `monPVCTemplate` and `monDataDirHostPath`.
+	if len(sc.Spec.MonDataDirHostPath) != 0 && sc.Spec.MonPVCTemplate != nil {
 		cephCluster.Spec.DataDirHostPath = sc.Spec.MonDataDirHostPath
-	}
-	if sc.Spec.MonPVCTemplate != nil {
 		cephCluster.Spec.Mon.VolumeClaimTemplate = sc.Spec.MonPVCTemplate
+		// If only the `monDataDirHostPath` is provided, the mons will be provisioned
+		// on the provided `monDataDirHostPath`.
+	} else if len(sc.Spec.MonDataDirHostPath) > 0 {
+		cephCluster.Spec.DataDirHostPath = sc.Spec.MonDataDirHostPath
+		// If only the `monPVCTemplate` is provided, the mons will provisioned on the
+		// provided `monPVCTemplate`.
+	} else if sc.Spec.MonPVCTemplate != nil {
+		cephCluster.Spec.Mon.VolumeClaimTemplate = sc.Spec.MonPVCTemplate
+		// If no `monPVCTemplate` and `monDataDirHostPath` is provided, the mons will
+		// be provisioned using the PVC template of first StorageDeviceSets if present.
+	} else if len(sc.Spec.StorageDeviceSets) > 0 {
```

Comment:
I think we should log a warning if none of these are matched.

Reply:
added it in the ensure function.